File: Data Structures & Algorithms/house-robber/submission-0.java
class Solution {
    public int rob(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, -1);
        return dfs(nums, 0, dp);
    }

    public int dfs(int[] nums, int i, int[] dp) {
        if (i >= nums.length) return 0;

        if (dp[i] != -1) {
            return dp[i];
        }

        dp[i] = Math.max(dfs(nums, i+1, dp), nums[i] + dfs(nums, i+2, dp));

        return dp[i];
    }
}

File: Data Structures & Algorithms/sliding-window-maximum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    maxSlidingWindow(nums, k) {
        const n = nums.length
        const res = Array(n - k + 1).fill(0)
        const window = Array(k).fill(0)
        for (let i = 0; i<k; i++) {
            window[i] = nums[i]
        }

        for(let i = k; i<n; i++) {
            res[i-k] = Math.max(...window)
            window[i % k] = nums[i]
        }

        res[n-k] = Math.max(...window)

        return res
    }
}
